### DISCUSSION TITLE
Are there many cases where data with wrapping characteristics need to be compared?

### DISCUSSION THREAD

**Discussion by Readon** [SpinalHDL/SpinalHDL#668] [Ideas]
In some cases, the number in the UInt is used to store the index, and when it overflows, the number is recalculated from 0.
When comparing on these UInts, < or > cannot be used directly due to the wrapping nature of the values.
If the distance between two UInts is always less than half of the maximum value, it is sufficient to check the sign bit of the result of subtracting two UInts.
Such as
``` scala
   val aSmaller = (a - b).msb
```
Importing a series of new operators like '<O' can make the use of these operations much easier.
The question is, is it necessary to add these operators?

As far as I know, the TCP protocol uses this mechanism.

**Comment by Dolu1990** [comment]
For operations which are "very specific" it may be better to avoid custom operator, but instead having a method which is named explicitly ?

to avoid confusion ^^

**Comment by Readon** [comment]
Alright, how about a wrapCmp function with comparison operator for UInt.
The usage would be
``` scala
val smaller = a.wrapCmp( <, b)
val bigger = a.wrapCmp( <=, b)
```
The first argument should be an operator.




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2022年3月28日(星期一) 下午5:09
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [SpinalHDL/SpinalHDL] Are there many cases where data with wrapping characteristics need to be compared? (Discussion #668)





 
For operations which are "very specific" it may be better to avoid custom operator, but instead having a method which is named explicitly ?
 
to avoid confusion ^^

**Reply by Dolu1990** [reply]
Hmm i'm realy not sure there is a way to have such syntax (<=, ) in scala

overall i kind of have often the issue of having usefull things in very specific cases, but hard to expose as it is so specific.
Another possibility is : 

```scala
class UInt extends BitVector{
  def wrapCmp = new {
     def <(v : UInt) : Bool
     def <=(v : UInt) : Bool
  }
}
a wrapCmp <= b
```

**Reply by Readon** [reply]
If I am using implicit function as this
``` scala
object UIntHelper {
  implicit class UIntWrap(val obj: UInt) {
    def wrap = new {
      def <(v: UInt): Bool = { (obj - v).msb }
      def >=(v: UInt): Bool = { !(<(v)) }
      def <=(v: UInt): Bool = { val result = obj - v; result === 0 || result.msb }
      def >(v: UInt): Bool = { !(<=(v)) }
    }
  }
}
```
"a.wrap < b" is working.

However, when I tried to add this to UInt class then obj should be instead by "this", while wrap is a new object where "this" is referring to wrap. That would lead to an error.
``` scala
def wrap = new {
      def <(v: UInt): Bool = { (this - v).msb } // here this is wrap, but not a UInt.
      def >=(v: UInt): Bool = { !(<(v)) }
      def <=(v: UInt): Bool = { val result = this - v; result === 0 || result.msb }
      def >(v: UInt): Bool = { !(<=(v)) }
    }
```

**Reply by Readon** [reply]
finally, I found that use _data but this could helps. 
define below in UInt class would work.
``` scala
  object wrap {
    def <(that: UInt): Bool = { (_data - that).msb }
    def >=(that: UInt): Bool = { !(<(that)) }
    def <=(that: UInt): Bool = { val result = _data - that; result === 0 || result.msb }
    def >(that: UInt): Bool = { !(<=(that)) }
  }
```

**Reply by Readon** [reply]
One more question is that where to put the tester for this logic.

**Comment by Readon** [comment]
there always be a balance on readability and generic. this would be a good solution.&nbsp;


one possible variation might be "a wrap <= b" where "<=" already stands for comparison.



---Original---
From: ***@***.***&gt;
Date: Wed, Mar 30, 2022 22:45 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [SpinalHDL/SpinalHDL] Are there many cases where data with wrapping characteristics need to be compared? (Discussion #668)




 
Hmm i'm realy not sure there is a way to have such syntax (<=, ) in scala
 
overall i kind of have often the issue of having usefull things in very specific cases, but hard to expose as it is so specific.
 Another possibility is :
 class UInt extends BitVector{   def wrapCmp = new {      def <(v : UInt) : Bool      def <=(v : UInt) : Bool   } } a wrapCmp <= b

**Reply by Readon** [reply]
I tried with implicit class in our project, it should be 
``` scala
  a.wrap <= b

  a wrap <= b //would lead a syntax error.
```

**Reply by Dolu1990** [reply]
Thanks for the test ^^

So, so far, i tend to think that the best maybe for now to just have a scala object with a few functions

**Reply by Readon** [reply]
> Thanks for the test ^^
> 
> So, so far, i tend to think that the best maybe for now to just have a scala object with a few functions

add a new object but functions for UInt？

**Reply by Readon** [reply]
I create a PR #675 to this problem.

**Reply by Dolu1990** [reply]
Thanks ^^
